### PR TITLE
Add support for a connection healthcheck

### LINKED ISSUE #274: Implement Connection.validate()
Re-enable `SpannerClientLibraryTestKit.validate()` to test.

V1 implementation is in `SpannerConnection.validate()` -- it relies on `GrpcClient.healthcheck`, which needs to be re-implemented relying on V2's SELECT statement execution functionality in `DatabaseClientReactiveAdapter`.

### PR DESCRIPTION
**Pull request by dzou**
This adds support for a Spanner connection health check.

Fixes #274 .

### CODE REVIEW

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryConnection.java [hunk 1]
```diff
@@ -134,7 +132,12 @@ public IsolationLevel getTransactionIsolationLevel() {
 
   @Override
   public Publisher<Boolean> validate(ValidationDepth depth) {
-    throw new UnsupportedOperationException();
+    if (depth == ValidationDepth.LOCAL) {
+      throw new UnsupportedOperationException(
+          "Local validation health checks is not supported for Cloud Spanner.");
```
Comment (dmitry-s) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/SpannerClientLibraryConnection.java:137]:
```suggestion
          "Local validation health checks are not supported for Cloud Spanner.");
```

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/DatabaseClientReactiveAdapter.java [hunk 2]
```diff
@@ -159,7 +161,31 @@ private void clearTransactionManager() {
...
+  }
+
+  /**
+   * Runs a health check query to determine if the connection is running.
+   *
+   * @return true if the connection is working, false if not.
+   */
+  public Mono<Boolean> healthCheck() {
+    return Mono.defer(() -> {
+      if (executorService.isShutdown()) {
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/DatabaseClientReactiveAdapter.java:174]: This seems like a good candidate for `ValidationDepth.LOCAL`.

I was thinking another useful local check would require saving `Spanner spannerClient` from the constructor, and calling `spannerClient.isClosed()`, but then realized we are not actually closing it, so I filed to track addressing the session leaking in #284.

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerClientLibraryTestKit.java [hunk 3]
```diff
@@ -304,9 +305,17 @@ public void columnMetadata() {
   }
 
   @Test
-  @Disabled // TODO: GH-274
+  @Override
```
Comment (elefeint) [inline cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerClientLibraryTestKit.java:308]: You might be able to just remove this override completely -- the `TestKit` parent should run as is.

**Review by elefeint — CHANGES_REQUESTED** [review]
Sorry for the delay!

Looks good, but we used to support local health check, and a part of your full healthcheck lends itself to becoming local.

This also led me into a rabbit hole, at the end of which was realizing we are not cleaning up client library-created sessions.

File: cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/DatabaseClientReactiveAdapter.java (hunk 2 above)
Comment (dzou) [reply cloud-spanner-r2dbc/src/main/java/com/google/cloud/spanner/r2dbc/v2/DatabaseClientReactiveAdapter.java]: Done.

File: cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerClientLibraryTestKit.java (hunk 3 above)
Comment (dzou) [reply cloud-spanner-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/it/SpannerClientLibraryTestKit.java]: Done.

**Review by elefeint — APPROVED** [review]
Thanks!

**Pull request merged**